fedimint-client/src/lib.rs

```diff
@@ -1375,67 +1380,78 @@ impl Client {
+            api: &DynGlobalApi,
+        ) -> (
+            PeerId,
+            Result<SupportedApiVersionsSummary, fedimint_api_client::api::PeerError>,
+        ) {
+            runtime::sleep(delay).await;
+            (
+                peer_id,
+                api.request_single_peer_typed::<SupportedApiVersionsSummary>(
+                    None,
+                    VERSION_ENDPOINT.to_owned(),
+                    ApiRequestErased::default(),
+                    peer_id,
+                )
+                .await,
             )
-            .await?)
-    }
-
-    pub async fn discover_common_api_version_static(
-        config: &ClientConfig,
-        client_module_init: &ClientModuleInitRegistry,
-        api: &DynGlobalApi,
-        mode: DiscoverCommonApiVersionMode,
-    ) -> anyhow::Result<ApiVersionSet> {
-        // We want to give it at least a good 30 second worth of trying, before we give
-        // up.
-        let deadline = now() + Duration::from_secs(30);
+        }
 
-        loop {
-            let res =
-                Self::discover_common_api_version_static_try(config, client_module_init, api, mode)
-                    .await;
+        for peer_id in num_peers.peer_ids() {
+            requests.push(make_request(Duration::ZERO, peer_id, &api))
+        }
 
-            if res.is_ok() {
-                return res;
-            }
+        let mut num_responses = 0;
 
-            if deadline < now() {
-                return res;
+        while let Some((peer_id, response)) = requests.next().await {
+            match response {
+                Err(err) => {
+                    if db
+                        .begin_transaction_nc()
+                        .await
+                        .get_value(&PeerLastApiVersionsSummaryKey(peer_id))
+                        .await
+                        .is_some()
+                    {
+                        debug!(target: LOG_CLIENT, %peer_id, %err, "Failed to refresh API versions of a peer, but we have a previous response");
+                    } else {
+                        debug!(target: LOG_CLIENT, %peer_id, %err, "Failed to refresh API versions of a peer, will retry");
+                        requests.push(make_request(Duration::from_secs(15), peer_id, &api))
+                    }
+                }
+                Ok(o) => {
+                    // Save the response to the database right away, just to
+                    // not lose it
+                    let mut dbtx = db.begin_transaction().await;
+                    dbtx.insert_entry(
+                        &PeerLastApiVersionsSummaryKey(peer_id),
+                        &PeerLastApiVersionsSummary(o),
+                    )
+                    .await;
+                    dbtx.commit_tx().await;
+                    num_responses += 1;
+                    // ignore errors: we don't care if anyone is still listening
+                    let _ = num_responses_sender.send(num_responses);
```

Comment:
nit: Instead of keeping `num_responses` around separately you could use [`send_modify`](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Sender.html#method.send_modify) directly on `num_responses_sender`